_release/bat/quotas_bat/quotas_bat_test.go

```diff
@@ -96,3 +97,235 @@ func TestQuotas(t *testing.T) {
+			_, err := bat.DeleteInstances(ctx, "", scheduled)
+			if err != nil {
+				t.Errorf("Failed to delete instances: %v", err)
+			}
+		}()
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Try launching instances with a quota.
+//
+// Sets a quota on the number of instances that can be launched and exceeds
+// that.
+//
+// Set a quota of two instances, "tenant-instances-quota", and then try and
+// start a set of three instances and check that the launch fails.
+func TestInstanceLimited(t *testing.T) {
+	ctx, cancelFunc := context.WithTimeout(context.Background(), standardTimeout)
+	defer cancelFunc()
+	tenants, err := bat.GetUserTenants(ctx)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if len(tenants) < 1 {
+		t.Fatal("Expected user to have access to at least one tenant")
+	}
+	tenantID := tenants[0].ID
+	origQuotas, err := bat.ListQuotas(ctx, "", "")
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = bat.UpdateQuota(ctx, "", tenantID, "tenant-instances-quota", "2")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	err = launchMultipleInstances(ctx, t, tenantID, 3)
+	if err == nil {
+		t.Errorf("Expected launch of instances to fail")
+	}
+
+	updatedQuotas, err := bat.ListQuotas(ctx, tenantID, "")
+	if err != nil {
+		t.Error(err)
+	}
+
+	err = restoreQuotas(ctx, tenantID, origQuotas, updatedQuotas)
+	if err != nil {
+		t.Error(err)
+	}
+}
+
+// Try launching instance that is over quota and check usage
+//
+// Sets a quota that would deny the instance from being created and
+// then check the usage reflects the instance failed to start.
+//
+// Set an memory quota limit < workload usage. Try and create instance
+// of workload and check that instance launch fails and that the usage
+// reflects that.
+func TestInstanceUsageAfterDenial(t *testing.T) {
+	ctx, cancelFunc := context.WithTimeout(context.Background(), standardTimeout)
+	defer cancelFunc()
+	tenants, err := bat.GetUserTenants(ctx)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if len(tenants) < 1 {
+		t.Fatal("Expected user to have access to at least one tenant")
+	}
+
+	tenantID := tenants[0].ID
+	origQuotas, err := bat.ListQuotas(ctx, "", "")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	err = bat.UpdateQuota(ctx, "", tenantID, "tenant-mem-quota", "0")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	wl, err := getContainerWorkload(ctx, tenantID)
+	if err != nil {
+		t.Skip()
+	}
+
+	instances, err := bat.LaunchInstances(ctx, "", wl.ID, 1)
+	if err == nil {
+		t.Error("Expected instance launch to be denied")
+	}
+
+	scheduled, err := bat.WaitForInstancesLaunch(ctx, "", instances, false)
+	if err != nil {
+		t.Fatalf("Instance %s did not launch: %v", instances[0], err)
+	}
+
+	updatedQuotas, err := bat.ListQuotas(ctx, tenantID, "")
+	if err != nil {
+		t.Error(err)
+	}
+
+	if !checkUsage(updatedQuotas, wl, 0) {
+		t.Error("Usage not recorded correctly")
+	}
+
+	err = restoreQuotas(ctx, tenantID, origQuotas, updatedQuotas)
+	if err != nil {
+		t.Error(err)
+	}
+
+	for _, i := range scheduled {
```

Comment:
Is there any reason not to just call DeleteInstances here?  Might speed things up a bit.